posts/2022-05-03-microstream-cdi-ext.adoc

```diff
@@ -0,0 +1,265 @@
+
+This is the _Root_ and _storage_ definition we have in the OpenLiberty example on GitHub.
+
+[source,Java]
+----
+@Storage
+public class Inventory {
+    private final Set<Product> products = new HashSet<>();
+
+    public void add(final Product product) {
+        Objects.requireNonNull(product, "product is required");
+        this.products.add(product);
+    }
+
+    public Set<Product> getProducts() {
+        return Collections.unmodifiableSet(this.products);
+    }
+
+    public Optional<Product> findById(final long id) {
+        return this.products.stream().filter(this.isIdEquals(id)).limit(1).findFirst();
+    }
+----
+
+By annotating it, we can have this object and the entire Object Graph, so including all the `Product`s we have gathered in the _Set_, marked as the _database_ that can be persisted. Besides the fact that we mark this class, we can implement all methods to perform operations on our inventory of products like adding, searching, updating, and deleting them.
+
+
+##Indicate Store actions
+
+The last thing that we need to do is indicate when we want to store the Object Graph on disk.  A CDI interceptor is ideal for that, and the CDI extension has defined the annotation `@Store` for this purpose.
+
+[source]
+----
+	@Inject
+	private Inventory inventory;
+
+
+	@Store
+	public Product save(final Product item)
+	{
+		this.inventory.add(item);
+		return item;
+	}
+----
+
+Whenever the method `save()` is executed, the interceptor makes sure the root instance, `Inventory` in our example since we have annotated it as such, is stored. The CDI extension tries to perform a few optimisations, but cannot exactly know what it needs to persist. For example, suppose we have the following Object Graph;
+
+Root
+-> Set<Person>; Person has reference to Address
+-> Set<Product>
+-> Set<Order>; Order has reference to Person and Product
+
+When you have a method `updateAddress(Person, Address)`, you should ideally only store only the Person as that is the only change. But we cannot indicate this requirement through annotations. If you have a very large Object Graph, it is recommended to inject the `StorageManager` and trigger the persistence of a single instance yourself.
+
+MicroStream can handle partial updates perfectly and works in a similar fashion as Git.  You can store the entire Object Graph and later on only the updated instances.  At startup, it assembles all pieces to reconstruct the latest situation when you stored something in the previous run.
+There is also continuous householding of the pieces going on so that old blobs are removed and others are reorganised to remove redundant info from the storage. You can configure the amount of time that is spent on this householding to balance the impact on the application throughput.
```

Comment:
```suggestion
There is also continuous housecleaning of the pieces going on so that old blobs are removed and others are reorganized to remove redundant info from the storage. You can configure the amount of time that is spent on this housecleaning to balance the impact on the application throughput.
```